esmtools/stats.py

```diff
@@ -135,317 +132,381 @@ def area_weight(da, area_coord="area"):
-            + "Please retry with a DataArray."
-        )
+    _check_y_not_independent_variable(y, dim)
+    x, slope_factor = _convert_time_and_return_slope_factor(x, dim)
 
-    x = da[dim]
-    # Linear regression doesn't like to work with actual datetime. So temporarily
-    # convert to integers.
-    # `np.object` covers cftime.
-    if np.issubdtype(x, np.datetime64) or isinstance(x, np.object):
-        x = np.arange(len(x))
-
-    if interpolate_na:
-        # borrowed from @ahuang11's implementation in `climpred`
-        da_dims_orig = list(da.dims)  # orig -> original
-        if len(da_dims_orig) > 1:
-            # want independent axis to be the leading dimension
-            da_dims_swap = da_dims_orig.copy()  # copy to prevent contamination
-
-            # https://stackoverflow.com/questions/1014523/
-            # simple-syntax-for-bringing-a-list-element-to-the-front-in-python
-            da_dims_swap.insert(0, da_dims_swap.pop(da_dims_swap.index(dim)))
-            da = da.transpose(*da_dims_swap)
-
-            # hide other dims into a single dim
-            da = da.stack({"other_dims": da_dims_swap[1:]})
-            dims_swapped = True
+    def _linear_slope(x, y, nan_policy):
+        if nan_policy in ["none", "propagate"]:
+            if has_missing(x) or has_missing(y):
+                return np.array([np.nan])
         else:
-            dims_swapped = False
-
-        # This is borrowed from @ahuang11's implementation in `climpred` to handle
-        # NaNs in the time series.
-        nan_locs = np.isnan(da.values)
-
-        # any(nan_locs.sum(axis=0)) fails if not 2D
-        if nan_locs.ndim == 1:
-            nan_locs = nan_locs.reshape(len(nan_locs), 1)
-
-        # check if there's any NaNs in the provided dim because
-        # interpolate_na is computationally expensive to run regardless of NaNs
-        # if nan_locs.sum(dim=dim).any():
-        if any(nan_locs.sum(axis=0)) > 0:
-            da = da.interpolate_na(dim)
-            if any(nan_locs[0, :]):
-                # [np.nan, 1, 2], no first value to interpolate from; back fill
-                da = da.bfill(dim)
-            if any(nan_locs[-1, :]):
-                # [0, 1, np.nan], no last value to interpolate from; forward fill
-                da = da.ffill(dim)
-
-        # this handles the other axes; doesn't matter since it won't affect the fit
-        da = da.fillna(0)
-        if dims_swapped:
-            da = da.unstack("other_dims").transpose(*da_dims_orig)
+            x, y = _handle_nans(x, y, nan_policy)
+        return np.polyfit(x, y, 1)[0]
 
-    results = xr.apply_ufunc(
-        lreg,
+    slopes = xr.apply_ufunc(
+        _linear_slope,
         x,
-        da,
-        input_core_dims=[[dim], [dim]],
-        output_core_dims=[[], [], [], [], []],
+        y,
+        nan_policy,
         vectorize=True,
-        dask="allowed",
+        dask="parallelized",
```

Comment:
Yes the speed is as good as we'll get it. We can only have `vectorize=False` if operations can be performed with the `axis` argument, i.e. over a dimension. Then we can just say `axis=-1`. `vectorize=True` makes it so that each grid cell is passed in individually so the operation is being done to individual time series. I don't see any way around this.